<a href="https://colab.research.google.com/github/BountyHunter1999/Machine-Learning-Zoom-Camp/blob/main/Homeworks/Homework%20%231.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing Modules

In [1]:
import numpy as np
import pandas as pd

# Getting Required Data

In [2]:
!wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv

--2021-09-09 09:41:33--  https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1475504 (1.4M) [text/plain]
Saving to: ‘data.csv.3’

data.csv.3          100%[===================>]   1.41M  --.-KB/s    in 0.02s   

2021-09-09 09:41:33 (60.3 MB/s) - ‘data.csv.3’ saved [1475504/1475504]



# Q1. Checking Numpy Version
[useful](https://stackoverflow.com/questions/1520234/how-do-i-check-which-version-of-numpy-im-using)




In [3]:
np.version

<module 'numpy.version' from '/usr/local/lib/python3.7/dist-packages/numpy/version.py'>

In [4]:
np.version.version

'1.19.5'

In [5]:
np.__version__

'1.19.5'

# Q2. Checking Pandas Version

In [6]:
pd.__version__

'1.1.5'

In [7]:
df = pd.read_csv("data.csv")

In [8]:
df.head()

Make       Model  Year  ... city mpg  Popularity   MSRP
0  BMW  1 Series M  2011  ...       19        3916  46135
1  BMW    1 Series  2011  ...       19        3916  40650
2  BMW    1 Series  2011  ...       20        3916  36350
3  BMW    1 Series  2011  ...       18        3916  29450
4  BMW    1 Series  2011  ...       18        3916  34500

[5 rows x 16 columns]

In [9]:
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [10]:
df['Make'] == "BMW"

0         True
1         True
2         True
3         True
4         True
         ...  
11909    False
11910    False
11911    False
11912    False
11913    False
Name: Make, Length: 11914, dtype: bool

In [11]:
bmw_data = df[df['Make'] == "BMW"]

In [12]:
bmw_data.describe()

Year   Engine HP  ...  Popularity           MSRP
count   334.000000  334.000000  ...       334.0     334.000000
mean   2014.347305  326.907186  ...      3916.0   61546.763473
std       4.387983  105.681659  ...         0.0   27982.682153
min    1995.000000  170.000000  ...      3916.0    4697.000000
25%    2015.000000  240.000000  ...      3916.0   42300.000000
50%    2016.000000  300.000000  ...      3916.0   51850.000000
75%    2016.000000  394.000000  ...      3916.0   79725.000000
max    2017.000000  600.000000  ...      3916.0  141200.000000

[8 rows x 8 columns]

# Question 3

## What's the average price of BMW cars in the dataset?
* 61546.763473




In [13]:
round(bmw_data.MSRP.mean(),2)

61546.76

# Question 4 
## Number of missing values in "Engine HP" after 2015

*   51



In [14]:
df["Engine HP"].isnull().sum()

69

In [15]:
af_2015 = df[df["Year"] >= 2015]

In [16]:
af_2015["Engine HP"].isnull().sum()

51

# Question 5 
## Does the mean change after filling missing values?

*   NO



In [17]:
af_2015_eng_hp =  af_2015["Engine HP"]

In [18]:
af_2015_eng_hp.isnull().sum()

51

In [19]:
mean_before = round(af_2015_eng_hp.mean())
mean_before

273

In [20]:
af_2015_eng_hp = af_2015_eng_hp.fillna(mean_before)

In [21]:
mean_after = round(af_2015_eng_hp.mean())
mean_after

273

# Question 6

## Steps:
1.   Select all the "Rolls-Royce" cars from the dataset.
2.   List item
3.   Select only columns "Engine HP", "Engine Cylinders", "highway MPG".
4.   Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 7 rows).
5.   Get the underlying NumPy array. Let's call it X.
6.   Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
7. Invert XTX.


## Sum of elements of the inverse matrix

*  0.0322

### 6.1 Select all the "Rolls-Royce" cars from the dataset.

In [22]:
rolls_royce = df[df["Make"] == "Rolls-Royce"]
len(rolls_royce)

31

### 6.2 List item

In [23]:
rolls_royce.head()

Make            Model  Year  ... city mpg  Popularity    MSRP
2921  Rolls-Royce         Corniche  2001  ...       10          86  359990
3505  Rolls-Royce             Dawn  2016  ...       12          86  335000
5275  Rolls-Royce  Ghost Series II  2015  ...       13          86  286750
5276  Rolls-Royce  Ghost Series II  2015  ...       13          86  319400
5277  Rolls-Royce  Ghost Series II  2016  ...       13          86  295850

[5 rows x 16 columns]

### 6.3 Select only columns "Engine HP", "Engine Cylinders", "highway MPG".

In [24]:
rolls_royce_3c = rolls_royce[[ "Engine HP", "Engine Cylinders", "highway MPG"]]

### 6.4 Drop all duplicated rows using drop_duplicates method

In [25]:
unique_rr3c = rolls_royce_3c.drop_duplicates()
len(unique_rr3c)

7

### 6.5 Get the underlying NumPy array

In [26]:
X = pd.DataFrame.to_numpy(unique_rr3c)
X.shape

(7, 3)

In [27]:
type(X)

numpy.ndarray

### 6.6 Compute matrix-matrix multiplication between the transpose of X and X

In [28]:
XTX = X.T.dot(X)
XTX

array([[1.754801e+06, 3.965600e+04, 6.519600e+04],
       [3.965600e+04, 9.280000e+02, 1.500000e+03],
       [6.519600e+04, 1.500000e+03, 2.454000e+03]])

### 6.7 Invert XTX.

In [29]:
XTX_INV = np.linalg.inv(XTX)
XTX_INV

array([[ 5.17815728e-05,  9.06587044e-04, -1.92984188e-03],
       [ 9.06587044e-04,  1.05723058e-01, -8.87084092e-02],
       [-1.92984188e-03, -8.87084092e-02,  1.05900809e-01]])

### 6.8 Sum of elements of the inverse matrix

In [30]:
round(XTX_INV.sum(),4)

0.0322

# Questions 7 
## STEPS:
1. Create an array y with values [1000, 1100, 900, 1200, 1000, 850, 1300].
2. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
3. What's the value of the first element of w?.
## Normal equation

* 



### 7.1 Create an array y with values [1000, 1100, 900, 1200, 1000, 850, 1300] 

In [31]:
y = np.array( [1000, 1100, 900, 1200, 1000, 850, 1300])
y

array([1000, 1100,  900, 1200, 1000,  850, 1300])

In [32]:
XTX_INV.shape

(3, 3)

In [33]:
X.T.shape

(3, 7)

In [34]:
X.shape

(7, 3)

In [35]:
y.shape

(7,)

### 7.2 Multiply the inverse of XTX with the transpose of X, and then multiply the result by y.

In [36]:
w = (XTX_INV @ X.T) @ y
w

array([ 0.19989598, 31.02612262, 31.65378877])

### 7.3 What's the value of the first element of w?.

In [37]:
round(w[0], 4)

0.1999